### PR TITLE
chore: Create v1beta1 APIs

### PR DESCRIPTION
**Pull request by jonathan-innis**
Fixes #N/A 

**Description**

- This change adds the first iteration of the `v1beta1` APIs for `NodeClass`
- These APIs carry similar validation mechanisms to the `v1alpha1` APIs with similar fields
- These APIs are still subject to change due to approval of #4274 

NOTE: This does not _release_ the APIs. This only preps other work that will lead up to the eventual formal release of these APIs into the project. Even after this PR is merged, these APIs _may_ change until the APIs are released through a formal release process of Karpenter.

**How was this change tested?**

`make presubmit`
`/karpenter snapshot`

**Does this change impact docs?**
- [ ] Yes, PR includes docs updates 
- [ ] Yes, issue opened: # 
- [x] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by jonathan-innis — COMMENTED** [review]
/karpenter snapshot

**Review by jonathan-innis — COMMENTED** [review]
/karpenter snapshot

File: pkg/apis/v1beta1/nodeclass.go [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+	// It must be in the appropriate format based on the AMIFamily in use. Karpenter will merge certain fields into
+	// this UserData to ensure nodes are being provisioned with the correct configuration.
+	// +optional
+	UserData *string `json:"userData,omitempty"`
+	// Role is the AWS identity that nodes use.
+	// +optional
+	Role *string `json:"role,omitempty"`
+	// SubnetSelectorTerms is a list of or subnet selector terms. The terms are ORed.
+	// +optional
+	SubnetSelectorTerms []SubnetSelectorTerm `json:"subnetSelectorTerms" hash:"ignore"`
```
Comment (ellistarn) [inline pkg/apis/v1beta1/nodeclass.go:41]: I kind of see this as akin to naming []foo as `fooSlice`. There are many types of selectors in k8s. 
```suggestion
	SubnetSelector []SubnetSelectorTerm `json:"subnetSelector" hash:"ignore"`
```

File: pkg/apis/v1beta1/nodeclass.go [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+	// SecurityGroupSelectorTerms is a list of or security group selector terms. The terms are ORed.
+	// +optional
+	SecurityGroupSelectorTerms []SecurityGroupSelectorTerm `json:"securityGroupSelectorTerms" hash:"ignore"`
+	// AMISelectorTerms is a list of or ami selector terms. The terms are ORed.
+	// +optional
+	AMISelectorTerms []AMISelectorTerm `json:"amiSelectorTerms,omitempty" hash:"ignore"`
```
Comment (ellistarn) [inline pkg/apis/v1beta1/nodeclass.go:47]: same

File: pkg/apis/v1beta1/nodeclass.go [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+	// Context is a Reserved field in EC2 APIs
+	// https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_CreateFleet.html
+	// +optional
+	Context *string `json:"context,omitempty"`
+	// Tags to be applied on ec2 resources like instances and launch templates.
+	// +optional
+	Tags map[string]string `json:"tags,omitempty"`
```
Comment (ellistarn) [inline pkg/apis/v1beta1/nodeclass.go:54]: How did we do the ordering of these? For folks reading, you could imagine hoisting this one higher up. Or we could alphabetize.

File: pkg/apis/v1beta1/nodeclass.go [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+	Status NodeClassStatus `json:"status,omitempty"`
+
+	// IsNodeTemplate tells Karpenter whether the in-memory representation of this object
+	// is actually referring to a AWSNodeTemplate object. This value is not actually part of the v1beta1 public-facing API
+	// TODO @joinnis: Remove this field when v1alpha5 is unsupported in a future version of Karpenter
+	IsNodeTemplate bool `json:"-" hash:"ignore"`
+}
+
+func (a *NodeClass) Hash() string {
+	hash, _ := hashstructure.Hash(a.Spec, hashstructure.FormatV2, &hashstructure.HashOptions{
```
Comment (ellistarn) [inline pkg/apis/v1beta1/nodeclass.go:263]: lo.must?

File: pkg/apis/v1beta1/nodeclass.go (hunk 1 above)
Comment (jonathan-innis) [reply pkg/apis/v1beta1/nodeclass.go:41]: Discussed offline. The consensus among us was that upstream K8s conveys ORed selectors through the word "terms"; therefore, we are going to propagate the word terms into our selector logic to convey the same thing.

File: pkg/apis/v1beta1/nodeclass.go [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+	UserData *string `json:"userData,omitempty"`
+	// Role is the AWS identity that nodes use.
+	// +optional
+	Role *string `json:"role,omitempty"`
+	// SubnetSelectorTerms is a list of or subnet selector terms. The terms are ORed.
+	// +optional
+	SubnetSelectorTerms []SubnetSelectorTerm `json:"subnetSelectorTerms" hash:"ignore"`
+	// SecurityGroupSelectorTerms is a list of or security group selector terms. The terms are ORed.
+	// +optional
+	SecurityGroupSelectorTerms []SecurityGroupSelectorTerm `json:"securityGroupSelectorTerms" hash:"ignore"`
```
[the remark this answers is not in the document]
Comment (jonathan-innis) [reply pkg/apis/v1beta1/nodeclass.go:44]: Commented above

File: pkg/apis/v1beta1/nodeclass.go (hunk 2 above)
Comment (jonathan-innis) [reply pkg/apis/v1beta1/nodeclass.go:47]: Commented above

File: pkg/apis/v1beta1/nodeclass.go (hunk 3 above)
Comment (jonathan-innis) [reply pkg/apis/v1beta1/nodeclass.go:54]: This is a good point. I think we want to do it by how common the fields are along with things that should be naturally grouped together. I don't think that alphabetic makes a lot of sense on its own.

**Review by ellistarn — APPROVED** [review]

**Pull request merged**